### PR TITLE
[2.11.0-beta1] experimental deferred request properties

### PR DESCRIPTION
**Pull request by MS-TimothyMothra**
Fix Issue #1173.
Second half of this fix.
This will use experimental config to enable deferred request properties.

### TODO
- Wait for BaseSDK 2.11.0-beta1 (contains Experimental)

### CODE REVIEW

File: Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs [hunk 1]
```diff
@@ -267,6 +267,13 @@ public void Initialize(TelemetryConfiguration configuration)
             this.telemetryClient = new TelemetryClient(configuration);
             this.telemetryClient.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("web:");
 
+            this.DisableTrackingProperties = configuration.EvaluateExperimentalFeature("deferRequestTrackingProperties");
```
Comment (lmolkova) [inline Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs]: > internal bool DisableTrackingProperties { get; set; } = false;

shouldn't they become private now?

Comment (MS-TimothyMothra) [reply Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs]: I think it had some tests that toggled this on/off.
tests require `internal`

File: Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs [hunk 2]
```diff
@@ -267,6 +267,13 @@ public void Initialize(TelemetryConfiguration configuration)
...
+            if (this.DisableTrackingProperties)
+            {
+                configuration.TelemetryProcessorChainBuilder.Use(next => new PostSamplingTelemetryProcessor(next));
```
Comment (cijothomas) [inline Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs]: can you validate that this achieves the intended effect? I believe this should be done on default sink's TP instead on common TP.

File: Src/PerformanceCollector/PerformanceCollector/Perf.csproj [hunk 3]
```diff
@@ -37,7 +37,7 @@
   
   <ItemGroup>
     <!--Common Dependencies-->
-    <PackageReference Include="Microsoft.ApplicationInsights" Version="2.10.0" />
+    <PackageReference Include="Microsoft.ApplicationInsights" Version="2.11.0-dev0-build32757" />
```
Comment (MS-TimothyMothra) [inline Src/PerformanceCollector/PerformanceCollector/Perf.csproj]: DO NOT MERGE.
Private build to test Experimental changes from BaseSDK

File: Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs (hunk 2 above)
Comment (MS-TimothyMothra) [reply Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs]: Done. wrote a bunch of new unit tests. :)

File: Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs [hunk 4]
```diff
@@ -114,6 +115,8 @@ public void Initialize(TelemetryConfiguration configuration)
             this telemetry processor will only collect for whichever instrumentation key is specified by the module in StartCollection call.
             */
 
+            this.EvaluateDisabledTrackingProperties = configuration.EvaluateExperimentalFeature("deferRequestTrackingProperties");
```
Comment (cijothomas) [inline Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: a line/comment to describe what this experimentalfeature would be really helpful.

File: Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs [hunk 5]
```diff
@@ -194,7 +197,7 @@ private ITelemetryDocument ConvertRequestToTelemetryDocument(RequestTelemetry re
             {
                 try
                 {
-                    // some of the requestTelemetry properties might be deffered by using RequestTrackingTelemetryModule.DisableTrackingProperties.
+                    // some of the requestTelemetry properties might be deferred by using RequestTrackingTelemetryModule.DisableTrackingProperties.
```
Comment (cijothomas) [inline Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: is this statement stil true in the context of new Experimental Feature suport?

Comment (cijothomas) [inline Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: also why is this appearing inside #if net45 only?

File: Src/PerformanceCollector/Perf.Shared.Tests/QuickPulse/QuickPulseTelemetryProcessorTests.cs [hunk 6]
```diff
@@ -2986,5 +2986,165 @@ public void QuickPulseTelemetryProcessorHandlesCalculatedMetricsInThreadSafeMann
...
+        public void VerifyInitializationWhenDeferredIsTrue()
+        {
+            var config = new TelemetryConfiguration();
+            config.ExperimentalFeatures.Add("deferRequestTrackingProperties");
+
+            var spy = new SimpleTelemetryProcessorSpy();
+            var telemetryProcessor = new QuickPulseTelemetryProcessor(spy);
+            telemetryProcessor.Initialize(config);
+
+            Assert.IsTrue(telemetryProcessor.EvaluateDisabledTrackingProperties);
```
Comment (cijothomas) [inline Src/PerformanceCollector/Perf.Shared.Tests/QuickPulse/QuickPulseTelemetryProcessorTests.cs]: why this test inside net45? The actual code is doing this irrespective of net45 or not.

File: Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs [hunk 7]
```diff
@@ -265,6 +266,13 @@ public void Initialize(TelemetryConfiguration configuration)
             this.telemetryClient = new TelemetryClient(configuration);
             this.telemetryClient.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("web:");
 
+            this.DisableTrackingProperties = configuration.EvaluateExperimentalFeature("deferRequestTrackingProperties");
```
Comment (cijothomas) [inline Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs]: "deferRequestTrackingProperties" - please move this to a const in shared proj. I believe its duplicated in qp and request tracking ..

File: Src/Web/Web.Shared.Net.Tests/RequestTrackingTelemetryModuleTest.cs [hunk 8]
```diff
@@ -478,6 +480,53 @@ public void TrackIntermediateRequestSetsProperties()
             Assert.True(intermediateRequest.Properties.ContainsKey("AI internal"));
         }
 
+        [TestMethod]
+        public void VerifyBehaviorWhenDeferredIsFalse()
+        {
+            var config = this.CreateDefaultConfig(HttpModuleHelper.GetFakeHttpContext());
+            var context = HttpModuleHelper.GetFakeHttpContext();
+
+            // Create and Validate RequestTrackingTelemetryModule
```
Comment (cijothomas) [inline Src/Web/Web.Shared.Net.Tests/RequestTrackingTelemetryModuleTest.cs]: Create, Initialize and Validate.

File: Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs (hunk 5 above)
Comment (MS-TimothyMothra) [reply Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: I think this is still valid.
The experimental feature is used to set this property, but this is still the property that decides to defer work.

Comment (MS-TimothyMothra) [reply Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: I believe this change was only intended for ASP.Net applications, not CORE.
As you can see I'm not changing that in my PR, that already exists.

File: Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs (hunk 4 above)
Comment (MS-TimothyMothra) [reply Src/PerformanceCollector/Perf.Shared/QuickPulseTelemetryProcessor.cs]: I will leave this comment on the new string constant.

File: Src/PerformanceCollector/Perf.Shared.Tests/QuickPulse/QuickPulseTelemetryProcessorTests.cs (hunk 6 above)
Comment (MS-TimothyMothra) [reply Src/PerformanceCollector/Perf.Shared.Tests/QuickPulse/QuickPulseTelemetryProcessorTests.cs]: the code path i'm testing is NET45 only.

Comment (MS-TimothyMothra) [reply Src/PerformanceCollector/Perf.Shared.Tests/QuickPulse/QuickPulseTelemetryProcessorTests.cs]: product code does have NET45 condition.

**Pull request merged**